### PR TITLE
Add logs cleanup mechanism

### PR DESCRIPTION
**Pull request by nyarian**
tari-project/wallet-android#318

### CODE REVIEW

**Comment by kukabi** [comment]
@nyarian I think it all boils down to just 33 lines and no additional libraries:
```kotlin
@Inject
@Named(WalletModule.FieldName.walletLogFilesDirPath)
lateinit var logFilesDirPath: String

private val maxTotalLogFilesBytes = 30 * 1024 * 1024

private val logFilesSizeCheckSubscription = Observable
    .timer(30, TimeUnit.MINUTES)
    .repeat()
    .subscribeOn(Schedulers.io())
    .observeOn(Schedulers.io())
    .subscribe {
        checkLogFilesSize()
    }

private fun List<File>.totalSize() = map { it.length() }.sum()

private fun checkLogFilesSize() {
    val logFiles = File(logFilesDirPath).listFiles()
        ?.asList()
        ?.filter { it.extension == "log" } // can be omitted
        ?.sortedBy { it.name }
        ?.toMutableList()
        ?: return
    while (logFiles.totalSize() >= maxTotalLogFilesBytes) {
        if (logFiles.size > 1) { // delete oldest log file
            logFiles[0].delete()
            logFiles.removeAt(0)
        } else { // empty current file
            // wipe the contents of the current log file
            logFiles[0].writeBytes(ByteArray(0))
        }
    }
}
;;;

**Comment by kukabi** [comment]
I think this task can be considered `WalletManager`'s responsibility, not sure if we need unit tests for this, simple enough to get away with just manual PoC imo.

**Comment by nyarian** [comment]
@kukabi I believe it can be reduced down to 20 lines or so if one is persistent enough

it boils down to the metrics of solution's quality. what are they? if it's only about lines count then the solution in the comment is clearly better. but let's consider other potential metrics as well:

- user service. the routine will run regardless of the service's lifecycle state. people tend to kill the application altogether when they are done with it by force stopping it from the system settings, and regular users that are not too deep into the cryptocurrency domain will do that even more often as they won't use Tari so much and they will know when they will get a TX, so the app will live as long as needed for a TX to arrive. android's workmanager provides an opportunity to guarantee that the routine will run. also it adds a bonus in constraints - it won't run if the app is low on battery, which is nice for the user as well

- abstraction & encapsulation. makes the system more understandable by hiding the implementation details away from the caller side and gives space for logic change. existing logic's codebase took a change in 3 lines: injected schedule and called `schedule.runDaily()`, so it can be said it pretty much implies a nice level of hiding the unnecessary details away and makes them easily substituted if a need arises. it makes the changes made easier for the system that's already in place. makes our customer pay less and take less risks if we are to change the functionality of the system

- unit tested. provides a physical metric - detects the inconsistencies introduced early at refactoring change (shrinks probability of regressions). as well as A&E, reduces risks, thus, reducing the maintenance costs

- integration tested. proof of work, not much to add. will fail as well if refactoring or reworking will break the stuff

- separation of concerns. activity is mostly a lifecycle events dispatcher, so it notifies the interested components about such events. all the logic concerning the schedule and filesystem interaction is separated, abstracted and encapsulated. by implementing it directly in service we are loosing the "user service" and "single responsibility", as the component will have one more thing to change (besides adding to the 1300 lines of code that are already there), so it makes maintenance cheaper as well

- failfast. if there is bug somewhere - we will detect it early at the development & QA stage, not letting the system behave in the undefined way

the solution's integration itself, aimed to satisfy these needs, took 134 lines, where changes to the existing codebase took 3 (in HomeActivity), so it's more "open-closed" principle oriented as well - system changed by behavior addition rather than the change

**Comment by kukabi** [comment]
* user service: If you make the interval short as something like 10 minutes, then it'll be pretty frequent and it will be extremely unlikely for the log files to exceed the limit. I think this is good enough. If it misses for some reason, it will be run immediately at the next startup, and it's fixed.

* abstraction & encapsulation: There's no end to abstraction & encapsulation, and sometimes a little fix is just good enough. In the end it's a matter of context and preferences. Too simple a fix to be taken that seriously imo.

* unit tests: I agree, we can have the simple solution in a class or something & unit test it. In general I'm really not sure if unit tests are of any value except for the major integration lines (it's FFI for us). Unit tests are also code and require maintenance. There's no end to testing too, but yes we can and maybe should have tests for it, adding that it's unlikely that this code is going to change once it's in place.

* separation of concerns: I propose having it in the `WalletManager`, it manages the wallet and related components, so this issue is 100% its responsibility. Plus, to mention again, it's a very minor fix, almost a hack, it really shouldn't introduce any new components.

* failfast: This short a snippet really shouldn't fail, absolutely not if you have unit tests for it on top.

* reusability: We shouldn't have new components just because there could be a use for them in the uncertain future. Same goes for interfaces: rather not have them if there's a single implementation and it's not being utilized in unit tests - classes also have interfaces.

So in summary I'd like to see the 20-line solution;) Or maybe slightly more sophisticated one that introduces 1 file for the class and 1 for unit tests. But really I'm fine with j. @nyarian just 20 lines. @nyarian

**Comment by nyarian** [comment]
@kukabi

the arguments shine when applied to the context, do you agree? let me be more specific

- user service. every 10 minutes will make impact on battery which we do have problems with already. and it does not consider "clear only if battery level is ok" case

- a&e & separation of concerns: for 20-lines solution there is no even start of abstraction. sometimes it is good enough, but given the context that walletservice is already 1300 lines long and already a big container of small fixes from here and there, won't another responsibility hurt it even harder? I believe they all start as 20-lines fix but aren't separated in future? speaking of walletmanager - manager is already a sign of scattered responsibility, oop rarely implies manager \ util \ helper objects as they do not fit the goals that the paradigm is aimed to solve: https://softwareengineering.stackexchange.com/a/129566/309254 , https://www.yegor256.com/2014/05/05/oop-alternative-to-utility-classes.html . basically, we don't know what the manager object does until we go under the hood, thus, the object by itself does not fit the paradigm structure

- unit tests provide costs reduction for the customer by detecting regressions early (it saves money by mitigating risks). if unit test refuses to compile when slight change is introduced to the tested class then that's a problem of unit test that tests the implementation details instead of pure input-output. so it's a design problem rather the one introduced by the notion unit-tests. integration tests are needed just to prove that system works. they have completely different purpose than unit tests

- failfast: it will fail once the changes are introduced. just a defense against regressions

- reusability: completely agree here, it was pointed out as an added bonus rather than a thing in mind that drove the design

if you insist on the solution after these clarifications are given, then, please, just make a note about it, and it will be implemented the way you insist on

**Comment by nyarian** [comment]
it was decided to go with walletmanager implementation
